snoop/html.py

```diff
@@ -0,0 +1,107 @@
+import re
+from bs4 import BeautifulSoup, UnicodeDammit
+import lxml.html
+import lxml.html.clean
+import lxml.etree
+
+
+def _extract_links(html):
+    from lxml.etree import fromstring, tostring, HTMLParser
+
+    tree = fromstring(html, parser=HTMLParser())
+
+    url_index = []
+    i = 0
+    for link in tree.findall('.//a'):
+        title = link.text
+        url = link.get('href')
+        if url:
+            i += 1
+            link.tag = 'span'
+            link.text = '%s [%s]' % (link.text, i)
+            url_index.append({'title': title, 'url': url})
+
+    html = tostring(tree, encoding="UTF-8", method='html').decode('utf-8')
+    print(type(html))
+    for i, link in enumerate(url_index):
+        if i == 0:
+            html += '<br/>\n<br/>\n'
+        html += '[%s] %s: \t %s <br/>\n' % (i + 1, link['title'], link['url'])
+    return html
+
+
+def _create_lxml_html_cleaner():
+    cleaner = lxml.html.clean.Cleaner()
+
+    # Only the tags will be removed, their content will get pulled up into
```

Comment:
Let's not copy/paste lxml documentation. Maybe [link](http://lxml.de/3.4/api/lxml.html.clean.Cleaner-class.html) to it?